File: python/algorithms/bubble.py
from util import swap_int

def bubble_sort(arr):
    n = len(arr)
    if n < 2:
        return
    for i in range(n - 1):
        swapped = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                swap_int(arr, j, j + 1)
                swapped = True
        if not swapped:
            break

File: c/algorithms/bubble.c
#include "bubble.h"
#include "../util.h"  // for swap_int

void bubble_sort(int *arr, size_t n) {
    if (!arr || n < 2) return;

    for (size_t i = 0; i + 1 < n; i++) {
        int swapped = 0;
        for (size_t j = 0; j + 1 < n - i; j++) {
            if (arr[j] > arr[j + 1]) {
                swap_int(&arr[j], &arr[j + 1]);
                swapped = 1;
            }
        }
        // Optimization trick - if no swaps occur, the array is already sorted and we can stop early.
        if (!swapped) break;
    }
}
